File: src/lib/api/git/types/repo.ts
// Repository identity, lifecycle, forge summary, and remotes — mirrors
// `src-tauri/src/git/types/repo.rs`.

import { isCommandError, type CommandError } from "@/lib/api/invoke";

export interface RepoSummary {
  path: string;
  workdir: string | null;
  headBranch: string | null;
  headOid: string | null;
  detached: boolean;
  /** True when HEAD is unborn (fresh `git init`, no commits yet) — the UI
   * shows "No commits yet" instead of "No branch". Optional for
   * backward-compatible fixtures; the backend always sends it. */
  unborn?: boolean;
  /** True when this checkout is a *linked* worktree. Optional for
   * backward-compatible fixtures; the backend always sends it. */
  isWorktree?: boolean;
  /** The main checkout's path for a linked worktree — the stable repository
   * identity (GL-109/GL-110); null for the main checkout itself. Optional for
   * fixtures; the backend always sends it. */
  mainPath?: string | null;
}

/** The one four-state reality the summary's head fields encode — a frontend
 * derivation, not a wire change: the raw fields stay on {@link RepoSummary}.
 * The backend guarantees the correlations (`head_branch` is null when
 * detached; an unborn HEAD's name is resolved from HEAD's symbolic target),
 * but each consumer used to re-derive the same `detached`/`unborn`/
 * `headBranch` ladder by hand — derive it once here and switch on `kind`.
 * `detached` wins even if a fixture somehow also sets `headBranch` (the wire
 * type permits it; the backend never does). */
export type HeadState =
  | { kind: "detached"; oid: string }
  | { kind: "unborn"; branch: string }
  | { kind: "branch"; branch: string; oid: string }
  | { kind: "none" };

/** Classify a summary's HEAD into the {@link HeadState} reality. Null-safe:
 * no repo is `none`, just like a summary whose head resolves to nothing.
 * Nullable raw fields are normalized to `""` (the backend always populates
 * them for the state that reads them). */
export function headStateOf(summary: RepoSummary | null): HeadState {
  if (!summary) return { kind: "none" };
  if (summary.detached) return { kind: "detached", oid: summary.headOid ?? "" };
  if (summary.unborn) return { kind: "unborn", branch: summary.headBranch ?? "" };
  if (summary.headBranch) {
    return { kind: "branch", branch: summary.headBranch, oid: summary.headOid ?? "" };
  }
  return { kind: "none" };
}

/** The `open_repo` rejections that mean the *path* is the problem, not the
 * read: a moved/deleted folder (`missingPath`) or one that lost its `.git`
 * (`notARepository`). The store swaps in the dedicated missing-repo state for
 * these instead of the raw libgit2 message (GL-108); every other kind keeps
 * the error-bar behavior. `path` is always populated for these kinds. */
export type RepoOpenError = CommandError & { kind: "missingPath" | "notARepository" };

/** Narrow an `api.openRepo` rejection to a {@link RepoOpenError}. */
export function isRepoOpenError(e: unknown): e is RepoOpenError {
  return isCommandError(e) && (e.kind === "missingPath" || e.kind === "notARepository");
}

/** Commit identity pinned for a repo: name + email, plus optional signing
 * config. Only the signing *reference* (GPG key id or SSH key path/literal) is
 * ever carried here — never a passphrase or private key. */
export interface RepoIdentity {
  name: string;
  email: string;
  /** `user.signingkey` pinned locally, if any. */
  signingKey?: string;
  /** `gpg.format` — "openpgp" or "ssh". */
  gpgFormat?: string;
  /** `commit.gpgsign` pinned locally, if any. */
  gpgSign?: boolean;
  /** `tag.gpgsign` pinned locally, if any. */
  tagGpgSign?: boolean;
}

/** The repo-identity snapshot a commit-creating write was composed against —
 * the tagged payload replacing the former `identity` + `identityCaptured` pair,
 * so "did not read one" and "read it; the repo had none" stay distinct. Build
 * it with `capturedIdentityArg` — the one place the distinction is decided. */
export type CapturedIdentity =
  | { mode: "notCaptured" }
  | { mode: "capturedNone" }
  | { mode: "card"; identity: RepoIdentity };

/** Optional signing config for {@link api.setRepoIdentity}. Tri-state per field
 * on the Rust side: omitted/`undefined` leaves the local key untouched, an
 * empty string unsets it, a value writes it. */
export interface RepoSigningConfig {
  signingKey?: string;
  gpgFormat?: string;
  gpgSign?: boolean;
  tagGpgSign?: boolean;
}

/** A signing key the user already has, for the profile editor's key picker.
 * Reference only — a full GPG fingerprint or SSH public-key path, never private
 * material. */
export interface SigningKey {
  /** Written to `user.signingkey` — a full GPG fingerprint or SSH public-key
   * path. */
  value: string;
  /** GPG uid, or SSH key type + comment. */
  label: string;
  format: "openpgp" | "ssh";
}

/** Presence + current branch of a previously-opened repo path (see Rust
 * `RecentStatus`). `exists: false` marks a path that no longer resolves on disk
 * so the onboarding list can flag it "Missing" (and session restore can drop
 * the tab). The tab strip shares this probe for worktree-tab labeling. */
export interface RecentStatus {
  path: string;
  exists: boolean;
  branch: string | null;
  /** True when the path is a *linked* worktree of some repository. Optional
   * for fixtures; the backend always sends it. */
  isWorktree?: boolean;
  /** The main checkout's path when `isWorktree` (see RepoSummary.mainPath).
   * Optional for fixtures; the backend always sends it. */
  mainPath?: string | null;
}

/** Payload of the `clone-progress` event streamed during a clone (see Rust
 * `CloneProgress`). `pct` is the blended overall completion 0–100. */
export interface CloneProgress {
  stage: string;
  pct: number;
}

/** Remote forge keys emitted by the backend's `ForgeKind::key()`
 * (`src-tauri/src/git/forge.rs`). This is the single source of truth on the TS
 * side — compare against `ForgeKind.GitHub` rather than a bare `"github"`
 * literal, so a typo fails to compile and a rename is one edit. Keep in sync
 * with the Rust enum across the IPC boundary. */
export const ForgeKind = {
  GitHub: "github",
  GitLab: "gitlab",
  Bitbucket: "bitbucket",
  AzureDevOps: "azure-devops",
  Gitea: "gitea",
  Forgejo: "forgejo",
  CursorOrigin: "cursor-origin",
} as const;
export type ForgeKind = (typeof ForgeKind)[keyof typeof ForgeKind];

/** Canonical Cursor Origin git host — keep in sync with Rust
 * `ForgeKind::CURSOR_ORIGIN_HOST`. */
export const CURSOR_ORIGIN_HOST = "origin.cursor.com";

/** Browser root for Cursor Origin PRs — keep in sync with Rust
 * `ForgeKind::CURSOR_ORIGIN_WEB_ROOT`. Distinct from the git host. */
export const CURSOR_ORIGIN_WEB_ROOT = "https://cursor.com/codebase";

/** Remote-forge summary driving the toolbar provider indicator. */
export interface RepoForge {
  /** True when the repo has at least one remote with a URL. */
  hasRemote: boolean;
  /** Forge key (see {@link ForgeKind}), or null when the host is unrecognised. */
  kind: ForgeKind | null;
  /** Human forge label ("GitHub", "GitLab", …), or null when unrecognised. */
  forge: string | null;
  /** Remote host (e.g. "github.com"), or null when no remote is configured. */
  host: string | null;
  /** Browser URL for the repo (`https://host/owner/repo`), or null when none. */
  webUrl: string | null;
}

/** A configured git remote (Repository settings → Remotes). */
export interface RemoteInfo {
  /** Remote name (e.g. "origin"). */
  name: string;
  /** Fetch URL. */
  fetchUrl: string;
  /** Push URL — equals the fetch URL unless a separate push URL is set. */
  pushUrl: string;
  /** True for the repo's default push remote. */
  isDefault: boolean;
}

File: src-tauri/src/git/types/repo.rs
//! Repository identity: the open-repo summary, open failures, recents, detected
//! forge, remotes, and the commit identity applied to a repo.

use serde::{Deserialize, Serialize};

/// High-level repository state shown in the title bar / status area.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoSummary {
    pub path: String,
    pub workdir: Option<String>,
    pub head_branch: Option<String>,
    pub head_oid: Option<String>,
    /// True when HEAD is detached (not on a branch).
    pub detached: bool,
    /// True when HEAD is unborn (fresh `git init`, no commits yet) — a real
    /// state, distinct from a read failure, so the UI can say "No commits yet"
    /// (GL-115).
    pub unborn: bool,
    /// True when this checkout is a *linked* worktree (not the main one).
    pub is_worktree: bool,
    /// The main checkout's path for a linked worktree — the stable repository
    /// identity that tab grouping and per-repo state key on (GL-109/GL-110).
    /// None for the main worktree itself (its own `path` is the identity).
    pub main_path: Option<String>,
}

/// Why `open_repo` failed, classified so the frontend can give a moved/deleted
/// repository its dedicated missing-repo state instead of the raw libgit2
/// message (GL-108). This is the one command with a structured IPC error —
/// everything else stays `Result<T, String>` (architecture-rules-rust §4); the
/// classification has to happen in Rust because only this side can distinguish
/// "path gone" from "not a repo" without matching on libgit2 message strings.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoOpenError {
    pub kind: RepoOpenErrorKind,
    /// Human-readable message (no libgit2 class/code jargon) for the error bar
    /// fallback when the failure isn't the missing-repo kind.
    pub message: String,
    /// The path the open was attempted with, echoed back for the missing state.
    pub path: String,
}

/// Classification of an [`RepoOpenError`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum RepoOpenErrorKind {
    /// The path no longer exists on disk — moved, deleted, or on an unmounted
    /// volume (so a retry after re-mounting is a real recovery path).
    Missing,
    /// The path exists but no longer resolves to a git repository.
    NotARepository,
    /// Any other open failure (permissions, corruption, …) — error-bar material.
    Other,
}

/// Presence + current branch of a previously-opened repository path, for the
/// onboarding "Recent" list and the tab strip's session-restore probe
/// (GL-109/GL-110 share this shape). `exists: false` flags a path that no
/// longer resolves on disk so the UI can mark it "Missing" / drop the tab.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentStatus {
    pub path: String,
    pub exists: bool,
    /// Currently checked-out branch (short name), or None when detached / gone.
    pub branch: Option<String>,
    /// True when the path is a *linked* worktree of some repository.
    pub is_worktree: bool,
    /// The main checkout's path when `is_worktree` (see [`RepoSummary`]).
    pub main_path: Option<String>,
}

/// Remote-forge summary for the toolbar provider indicator. Pure libgit2
/// detection from the configured remote URLs — no network or auth probing.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoForge {
    /// True when the repo has at least one remote with a non-empty URL.
    pub has_remote: bool,
    /// Lowercase forge key ("github", "gitlab", …). None when there is no
    /// remote or the host is not a recognised forge.
    pub kind: Option<String>,
    /// Human-readable forge label ("GitHub", "GitLab", …). None when unclassified.
    pub forge: Option<String>,
    /// Remote host (e.g. "github.com"). None when no remote URL is configured.
    pub host: Option<String>,
    /// Browser URL for the repo (`https://host/owner/repo`), derived from the
    /// remote URL. None when no path can be parsed (e.g. no remote).
    pub web_url: Option<String>,
}

/// A single configured git remote, for the Repository settings → Remotes panel.
/// Pure libgit2 read of `.git/config`; provider classification of the URL is
/// done on the frontend (shared with the add/edit validation).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteInfo {
    /// Remote name (e.g. "origin").
    pub name: String,
    /// Fetch URL.
    pub fetch_url: String,
    /// Push URL — equals the fetch URL unless a separate `pushurl` is set.
    pub push_url: String,
    /// True for the repo's default push remote (the current branch's upstream
    /// remote, else "origin", else the first remote).
    pub is_default: bool,
}

/// The commit identity pinned in a repo's *local* git config (`user.name` /
/// `user.email`). `None` from the read side means nothing is pinned locally and
/// the repo defers to global git config.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoIdentity {
    pub name: String,
    pub email: String,
    /// `user.signingkey` pinned locally — a GPG key id or an SSH key path /
    /// literal. The passphrase / private key never lives here: only the
    /// reference, unlocked at use time by ssh-agent / gpg-agent / the OS
    /// keychain. Omitted from JSON when unset.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signing_key: Option<String>,
    /// `gpg.format` — "openpgp" or "ssh" — pinned locally, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gpg_format: Option<String>,
    /// `commit.gpgsign` pinned locally, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gpg_sign: Option<bool>,
    /// `tag.gpgsign` pinned locally, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag_gpg_sign: Option<bool>,
}

/// The repo-identity snapshot a commit-creating write was composed against —
/// crosses IPC in place of the former `identity` + `identityCaptured` pair, so
/// "did not read the identity" (NotCaptured) and "read it; the repo had none"
/// (CapturedNone) are distinct variants rather than a representable-invalid
/// `captured: false` + `identity: Some(card)` combination. The per-commit
/// author pin (`name`/`email`) is a separate, independent argument.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(tag = "mode", rename_all = "camelCase")]
pub enum CapturedIdentity {
    /// The composer had not read the repo identity when the write was queued.
    NotCaptured,
    /// The repo had no identity card when the write was composed — "this
    /// computer". A card landing afterwards still fails closed.
    CapturedNone,
    /// The repo had this card. Any field changing before the write runs fails
    /// closed.
    Card { identity: RepoIdentity },
}

/// A signing key the user already has, offered by the profile editor's key
/// picker. References only — a full GPG fingerprint or an SSH public-key path —
/// never private key material or a passphrase.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SigningKey {
    /// The value written to `user.signingkey`: a full GPG fingerprint, or an SSH
    /// public-key path.
    pub value: String,
    /// Human-readable label — the GPG uid, or the SSH key type + comment.
    pub label: String,
    /// `gpg.format` for this key: "openpgp" or "ssh".
    pub format: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn captured_identity_deserializes_the_wire_shapes() {
        // The inbound IPC payload replacing the former identity +
        // identityCaptured pair; each variant's exact JSON is pinned so the
        // frontend helper and this enum cannot drift apart.
        assert_eq!(
            serde_json::from_value::<CapturedIdentity>(serde_json::json!({"mode": "notCaptured"}))
                .unwrap(),
            CapturedIdentity::NotCaptured
        );
        assert_eq!(
            serde_json::from_value::<CapturedIdentity>(serde_json::json!({"mode": "capturedNone"}))
                .unwrap(),
            CapturedIdentity::CapturedNone
        );
        assert_eq!(
            serde_json::from_value::<CapturedIdentity>(serde_json::json!({
                "mode": "card",
                "identity": {"name": "Ada", "email": "ada@example.test"}
            }))
            .unwrap(),
            CapturedIdentity::Card {
                identity: RepoIdentity {
                    name: "Ada".to_string(),
                    email: "ada@example.test".to_string(),
                    signing_key: None,
                    gpg_format: None,
                    gpg_sign: None,
                    tag_gpg_sign: None,
                }
            }
        );
    }

    #[test]
    fn captured_identity_rejects_the_legacy_flat_shape() {
        // The old wire sent bare identity/identityCaptured keys — including
        // the contradictory "not captured" + card combination that routed
        // through the weaker stale-card guard. Without a `mode` tag those
        // payloads now fail to deserialize instead of silently degrading.
        assert!(
            serde_json::from_value::<CapturedIdentity>(serde_json::json!({
                "identity": {"name": "Ada", "email": "ada@example.test"},
                "identityCaptured": false
            }))
            .is_err()
        );
        assert!(
            serde_json::from_value::<CapturedIdentity>(serde_json::json!({
                "identity": null,
                "identityCaptured": true
            }))
            .is_err()
        );
    }
}
